src/Microsoft.DotNet.Build.Tasks/IsRestoreRequired.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#define read
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using NuGet.Common;
+using NuGet.LibraryModel;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.ProjectModel;
+using NuGet.Versioning;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public partial class IsRestoreRequired : Task
+    {
+        [Required]
+        public ITaskItem[] ProjectJsons { get; set; }
+
+        [Required]
+        public string PackagesFolder { get; set; }
+
+
+        [Output]
+        public bool RestoreRequired { get; set; }
+
+        [Output]
+        public ITaskItem[] ProjectJsonsRequiringRestore { get; set; }
+
+        public override bool Execute()
+        {
+            var packagesChecked = new HashSet<PackageIdentity>();
+            var packageResolver = new VersionFolderPathResolver(PackagesFolder);
+            List<ITaskItem> needRestore = new List<ITaskItem>();
+            var lockFileFormat = new LockFileFormat();
+
+            ProjectJsons.AsParallel().ForAll(project =>
+            {
+                string projectJsonPath = project.GetMetadata("FullPath");
+                string projectLockJsonPath = ProjectJsonPathUtilities.GetLockFilePath(projectJsonPath);
+
+                if (!File.Exists(projectLockJsonPath))
+                {
+                    Log.LogMessage(MessageImportance.Low, $"{projectJsonPath} requires restore because {projectLockJsonPath} is missing.");
+                    needRestore.Add(project);
```

Comment:
List is not thread safe so we run the risk of race conditions when we these adds in parallel.